tools/crm_attribute.c

```diff
@@ -30,357 +30,310 @@
-        "\tDelete the attribute/option", pcmk__option_default
-    },
-    {
-        "-spacer-", no_argument, NULL, '-',
-        "\nAdditional Options:", pcmk__option_default
-    },
-    {
-        "node", required_argument, NULL, 'N',
-        "Set a node attribute for named node (instead of a cluster option). "
-            "See also: -l",
-        pcmk__option_default
-    },
-    {
-        "type", required_argument, NULL, 't',
-        "Which part of the configuration to update/delete/query the option in",
-        pcmk__option_default
-    },
-    {
-        "-spacer-", no_argument, NULL, '-',
-        "\t\t\tValid values: crm_config, rsc_defaults, op_defaults, tickets",
-        pcmk__option_default
-    },
-    {
-        "lifetime", required_argument, NULL, 'l',
-        "Lifetime of the node attribute", pcmk__option_default
-    },
-    {
-        "-spacer-", no_argument, NULL, '-',
-        "\t\t\tValid values: reboot, forever", pcmk__option_default
-    },
-    {
-        "utilization", no_argument, NULL, 'z',
-        "Set an utilization attribute for the node.", pcmk__option_default
-    },
-    {
-        "default", required_argument, NULL, 'd',
-        "(Advanced) Default value to display if none is found in configuration",
-        pcmk__option_default
-    },
-    {
-        "inhibit-policy-engine", no_argument, NULL, '!',
-        NULL, pcmk__option_hidden
-    },
+    if (options.attr_value) {
+        free(options.attr_value);
+    }
+
+    options.attr_value = NULL;
+    return TRUE;
+}
+
+static gboolean
+promotion_cb(const gchar *option_name, const gchar *optarg, gpointer data, GError **error) {
+    options.promotion_score = TRUE;
+
+    if (options.attr_name) {
+        free(options.attr_name);
+    }
+
+    options.attr_name = strdup(pcmk_promotion_score_name(optarg));
+    return TRUE;
+}
+
+static gboolean
+update_cb(const gchar *option_name, const gchar *optarg, gpointer data, GError **error) {
+    options.command = 'v';
+
+    if (options.attr_value) {
+        free(options.attr_value);
+    }
+
+    options.attr_value = strdup(optarg);
+    return TRUE;
+}
+
+static gboolean
+utilization_cb(const gchar *option_name, const gchar *optarg, gpointer data, GError **error) {
+    if (options.type) {
+        free(options.type);
+    }
+
+    options.type = strdup(XML_CIB_TAG_NODES);
+
+    if (options.set_type) {
+        free(options.set_type);
+    }
+
+    options.set_type = strdup(XML_TAG_UTILIZATION);
+    return TRUE;
+}
+
+static gboolean
+value_cb(const gchar *option_name, const gchar *optarg, gpointer data, GError **error) {
+    options.command = 'G';
+
+    if (options.attr_value) {
+        free(options.attr_value);
+    }
+
+    options.attr_value = NULL;
+    return TRUE;
+}
 
-    /* legacy */
-    {
-        "quiet", no_argument, NULL, 'Q',
-        NULL, pcmk__option_hidden
+static GOptionEntry selecting_entries[] = {
+    { "id", 'i', 0, G_OPTION_ARG_STRING, &options.attr_id,
```

Comment:
If you're populating strings directly with glib, you have to g_free() them

Reply:
Yeah, one day I'll actually remember this.